idp-binding-service/src/main/java/io/mosip/idp/binding/services/KeyBindingServiceImpl.java

```diff
@@ -0,0 +1,320 @@
+		} catch (KeyBindingException e) {
+			log.error("Failed to bind the key : {}", walletBindingRequest.getTransactionId(), e);
+			throw new IdPException(e.getErrorCode());
+		}
+
+		if (keyBindingResult == null || keyBindingResult.getCertificate() == null || keyBindingResult.getPartnerSpecificToken() == null) {
+			log.error("wallet binding failed with result : {}", keyBindingResult);
+			throw new IdPException(KEY_BINDING_FAILED);
+		}
+
+		PublicKeyRegistry publicKeyRegistry = storeKeyBindingDetailsInRegistry(walletBindingRequest.getIndividualId(),
+				keyBindingResult.getPartnerSpecificToken(), publicKey, keyBindingResult.getCertificate(),
+				walletBindingRequest.getAuthFactorTypes());
+
+		return new WalletBindingResponse(walletBindingRequest.getTransactionId(),
+				encryptBindingId ? getJWE(walletBindingRequest.getPublicKey(), publicKeyRegistry.getWalletBindingId()) :
+						publicKeyRegistry.getWalletBindingId(),
+				keyBindingResult.getCertificate(),
+				publicKeyRegistry.getExpiredtimes().format(DateTimeFormatter.ofPattern(UTC_DATETIME_PATTERN)));
+	}
+
+	@Override
+	public ValidateBindingResponse validateBinding(ValidateBindingRequest validateBindingRequest) throws IdPException {
+		String individualIdHash = getIndividualIdHash(validateBindingRequest.getIndividualId());
+		Optional<PublicKeyRegistry> publicKeyRegistry = publicKeyRegistryRepository.findByIdHashAndExpiredtimesGreaterThan(individualIdHash,
+				LocalDateTime.now(ZoneOffset.UTC));
+		if (!publicKeyRegistry.isPresent())
+			throw new IdPException(ErrorConstants.KEY_BINDING_NOT_FOUND);
+
+		List<String> supportedAuthFactors = null;
+		try {
+			supportedAuthFactors = objectMapper.readValue(publicKeyRegistry.get().getAuthFactors(), new TypeReference<List<String>>() {});
+		} catch (IllegalArgumentException | JsonProcessingException e) {
+			log.error("Failed to parse supported auth-factors", e);
+		}
+
+		//check if provided challenge auth-factor is the bound auth-factor-type for the provided individualId
+		if(CollectionUtils.isEmpty(supportedAuthFactors) || !supportedAuthFactors.contains(validateBindingRequest.getChallenge().getAuthFactorType()))
+			throw new IdPException(ErrorConstants.UNBOUND_AUTH_FACTOR);
+
+		X509Certificate x509Certificate = null;
+		try {
+			x509Certificate = (X509Certificate) keymanagerUtil.convertToCertificate(publicKeyRegistry.get().getCertificate());
+			//TODO should we validate the certificate?
+		} catch (KeymanagerServiceException e) {
+			log.error("Certificate parsing failed", e);
+			throw new IdPException(ErrorConstants.INVALID_CERTIFICATE);
+		}
+
+		switch (validateBindingRequest.getChallenge().getAuthFactorType()) {
+			case "WLA" :
+				verifyWLAToken(validateBindingRequest.getIndividualId(), validateBindingRequest.getChallenge().getChallenge(),
+						validateBindingRequest.getChallenge().getFormat(), publicKeyRegistry.get().getPublicKey(),
+						X509Util.x5t(x509Certificate));
```

Comment:
Can we store this thumbprint while binding, so we don't compute this everytime ?

Reply:
Based on the discussion, store certificate thumbprint hash.

Reply:
I don't see this change of storing the thumbprint

Reply:
it is not implemented, will be taking up in next PR.